sklearn/mixture/tests/test_gmm.py

```diff
@@ -334,6 +334,73 @@ def test_aic():
         assert_true(np.abs(g.bic(X) - bic) / n_samples < bound)
 
 
+def _test_positive_definite_covars(covariance_type):
+    """ Test that covariance matrices do not become non positive definite
+
+    Due to the accumulation of round-off errors, the computation of the
+    covariance  matrices during the learning phase could lead to non-positive
+    definite covariance matrices. Namely the use of the formula:
+
+    .. math:: C = (\\sum_i w_i \\boldsymbol x_i \\boldsymbol x_i^T)
+              - \\boldsymbol \\mu \\boldsymbol \\mu^T
+
+    instead of:
+
+    .. math:: C = \\sum_i w_i (\\boldsymbol x_i - \\boldsymbol \\mu)
+                  (\\boldsymbol x_i - \\boldsymbol \\mu)^T
+
+    while mathematically equivalent, was observed a ``LinAlgError`` exception,
+    when computing a ``GMM`` with full covariance matrices and fixed mean.
+
+    This function ensures that some later optimization will not introduce the
+    problem again.
+"""
+    rng = np.random.RandomState(1)
+    # we build a dataset with 2 2d component. The components are unbalanced
+    # (respective weights 0.9 and 0.1)
+    X = rng.randn(100, 2)
+    X[-10:] += (3, 3)  # Shift the 10 last points
+
+    gmm = mixture.GMM(2, params="wc", covariance_type=covariance_type,
+                      min_covar=1e-3)
+
+    # The following line can raise an exception with  full covariance matrices
+    # due to the use of the Cholesky decomposition when computing the density.
+    gmm.fit(X)
+
+    if covariance_type == "diag" or covariance_type == "spherical":
+        assert(gmm.covars_.min() > 0)
```

Comment:
Actually an even better test would be to check:

```python
assert_greater(np.linalg.det(gmm.covars_), 0)
```

Otherwise you just test for positive semi-definiteness which is not what the name of the function says. I agree this is nitpicking ;)